code/workspaces/infrastructure-api/src/stacks/shareStackManager.spec.js

```diff
@@ -0,0 +1,222 @@
+import { stackTypes } from 'common';
+import { NOTEBOOK_CATEGORY, SITE_CATEGORY } from 'common/src/config/images';
+import { handleSharedChange } from './shareStackManager';
+import config from '../config/config';
+import secretManager from '../credentials/secretManager';
+import jobApi from '../kubernetes/jobApi';
+import ingressApi from '../kubernetes/ingressApi';
+import deploymentApi from '../kubernetes/deploymentApi';
+import zeppelin from './zeppelinStack';
+
+const projectKey = 'project';
+const name = 'stackname';
+
+const volumeMount = 'volume';
+
+const { JUPYTERLAB, ZEPPELIN, RSTUDIO, RSHINY } = stackTypes;
+
+const getParams = () => ({
+  projectKey,
+  name,
+});
+const getExisting = () => ({
+  shared: 'project',
+  visible: 'project',
+  volumeMount,
+});
+
+// Function to help 'expect' numbers of calls for certain functions (so we don't need to replicate this several times).
+const expectCalls = ({
+  createJob = 0,
+  patchIngress = 0,
+  createStackCredentialSecret = 0,
+  getDeployment = 0,
+  restartDeployment = 0,
+  generateNewShiroIni = 0,
+}) => {
+  expect(jobApi.createJob).toHaveBeenCalledTimes(createJob);
+  expect(ingressApi.patchIngress).toHaveBeenCalledTimes(patchIngress);
+  expect(secretManager.createStackCredentialSecret).toHaveBeenCalledTimes(createStackCredentialSecret);
+  expect(deploymentApi.getDeployment).toHaveBeenCalledTimes(getDeployment);
+  expect(deploymentApi.restartDeployment).toHaveBeenCalledTimes(restartDeployment);
+  expect(zeppelin.generateNewShiroIni).toHaveBeenCalledTimes(generateNewShiroIni);
+};
+
+describe('handleSharedChange', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    jest.mock('../credentials/secretManager');
+    jest.mock('../kubernetes/jobApi');
+    jest.mock('../kubernetes/ingressApi');
+    jest.mock('../kubernetes/deploymentApi');
+    jest.mock('./zeppelinStack');
+
+    secretManager.createStackCredentialSecret = jest.fn();
```

Comment:
I can't quite work out what's not working here but it feels like this shouldn't be needed if you have the `jest.mock('../credentials/secretmanager');` at the top after the imports. This should set up a `jest.fn()` on each method - it might need a `resetAllMocks` instead of a `clearAllMocks`.

Reply:
I can't remember what wasn't working before now, but I've updated it and it seems to be fine.

Reply:
Looks good